File: surf-redis-api/src/main/java/dev/slne/surf/redis/codec/RedisVarLong.java
package dev.slne.surf.redis.codec;

import io.netty.buffer.ByteBuf;
import io.netty.handler.codec.DecoderException;

public final class RedisVarLong {

    public static final int MAX_VARLONG_SIZE = 10;
    private static final int DATA_BITS_MASK = 127;
    private static final int CONTINUATION_BIT_MASK = 128;
    private static final int DATA_BITS_PER_BYTE = 7;

    private RedisVarLong() {
        throw new AssertionError("No instances");
    }

    public static int getByteSize(final long value) {
        for (int i = 1; i < MAX_VARLONG_SIZE; i++) {
            if ((value & -1L << i * DATA_BITS_PER_BYTE) == 0L) {
                return i;
            }
        }

        return MAX_VARLONG_SIZE;
    }

    public static boolean hasContinuationBit(final byte in) {
        return (in & CONTINUATION_BIT_MASK) == CONTINUATION_BIT_MASK;
    }

    public static long read(final ByteBuf input) {
        long out = 0L;
        int bytes = 0;

        byte in;
        do {
            in = input.readByte();
            out |= (long) (in & (byte) DATA_BITS_MASK) << bytes++ * DATA_BITS_PER_BYTE;
            if (bytes > MAX_VARLONG_SIZE) {
                throw new DecoderException("RedisVarLong too big");
            }
        } while (hasContinuationBit(in));

        return out;
    }

    public static ByteBuf write(final ByteBuf output, long value) {
        while ((value & -(long) CONTINUATION_BIT_MASK) != 0L) {
            output.writeByte((int) (value & DATA_BITS_MASK) | CONTINUATION_BIT_MASK);
            value >>>= DATA_BITS_PER_BYTE;
        }

        output.writeByte((int) value);
        return output;
    }
}

File: surf-redis-api/src/main/kotlin/dev/slne/surf/redis/codec/RedisVarLong.kt
package dev.slne.surf.redis.codec

import io.netty.buffer.ByteBuf

object RedisVarLong {

    const val MAX_VARLONG_SIZE: Int = 10

    private const val DATA_BITS_MASK: Long = 0x7F
    private const val CONTINUATION_BIT_MASK: Int = 0x80
    private const val DATA_BITS_PER_BYTE: Int = 7

    @JvmStatic
    fun getByteSize(value: Long): Int {
        for (i in 1 until MAX_VARLONG_SIZE) {
            if ((value and (-1L shl (i * DATA_BITS_PER_BYTE))) == 0L) {
                return i
            }
        }

        return MAX_VARLONG_SIZE
    }

    @JvmStatic
    fun hasContinuationBit(input: Byte): Boolean {
        return (input.toInt() and CONTINUATION_BIT_MASK) == CONTINUATION_BIT_MASK
    }

    @JvmStatic
    fun read(input: ByteBuf): Long {
        var out = 0L
        var bytes = 0

        var byte: Byte
        do {
            byte = input.readByte()

            out = out or ((byte.toLong() and DATA_BITS_MASK) shl (bytes++ * DATA_BITS_PER_BYTE))

            checkDecoding(bytes <= MAX_VARLONG_SIZE) { "RedisVarLong too big" }
        } while (hasContinuationBit(byte))

        return out
    }

    @JvmStatic
    fun write(output: ByteBuf, value: Long): ByteBuf {
        var remaining = value

        while ((remaining and DATA_BITS_MASK.inv()) != 0L) {
            output.writeByte(((remaining and DATA_BITS_MASK).toInt()) or CONTINUATION_BIT_MASK)
            remaining = remaining ushr DATA_BITS_PER_BYTE
        }

        output.writeByte(remaining.toInt())
        return output
    }
}